candle-core/tests/quantized_tests.rs

```diff
@@ -124,51 +124,162 @@ fn quantize_q4_0() -> Result<()> {
+    assert!(
+        size % crate::quantized::k_quants::QK_K == 0,
+        "size must be a multiple of {}",
+        crate::quantized::k_quants::QK_K
+    );
 
-    let src = (0..256 * 4)
-        .map(|v| (v as f32 - 512.) / 1024.)
+    let src = (0..size)
+        .map(|v| (v as f32 - size as f32 / 2.) * bound / (size as f32 / 2.))
         .collect::<Vec<_>>();
-    let mut dst = vec![0f32; 256 * 4];
-    let mut quant = vec![BlockQ8K::zeros(); 4];
-    BlockQ8K::from_float(&src, &mut quant)?;
-    BlockQ8K::to_float(&quant, dst.as_mut_slice())?;
+
+    let dst = vec![0f32; size];
+    assert_eq!([src[0], src[size / 2]], [-bound, 0.0]);
+    (src, dst)
+}
+
+/// Round a vector
+fn round_vector(values: &[f32]) -> Vec<f32> {
+    values
+        .iter()
+        .map(|x| (1000. * x).round() / 1000.)
+        .collect::<Vec<_>>()
+}
+
+fn compare_with_error(values: &[f32], expected: &[f32], precision: usize, tolerance: f32) {
+    let factor = 10f32.powi(precision as i32);
+    for (i, (value, expected_value)) in values.iter().zip(expected.iter()).enumerate() {
+        let rounded_value = (value * factor).round() / factor;
+        let rounded_expected_value = (expected_value * factor).round() / factor;
+        let difference = (rounded_value - rounded_expected_value).abs();
+
+        assert!(
+            difference < tolerance,
+            "Error at index {}: value = {}, expected = {}. Difference = {} exceeds tolerance = {}.",
+            i,
+            rounded_value,
+            rounded_expected_value,
+            difference,
+            tolerance
+        );
+    }
+}
+
+fn quantize_roundtrip<T: GgmlType>(src: &[f32], dst: &mut [f32]) -> Result<Vec<T>> {
+    let mut quant = vec![T::zeros(); src.len() / T::BLCK_SIZE];
+    T::from_float(src, &mut quant)?;
+    T::to_float(&quant, dst)?;
+    Ok(quant)
+}
+
+#[test]
+fn quantize_q2k() -> Result<()> {
+    use k_quants::BlockQ2K;
+
+    let (src, mut dst) = get_test_vector(0.5, Some(1024));
+    let _quant = quantize_roundtrip::<BlockQ2K>(src.as_slice(), dst.as_mut_slice())?;
+    compare_with_error(src.as_slice(), dst.as_slice(), 6, 0.1);
+
+    // Test some specific values
+    dst = round_vector(&dst);
+    assert_eq!(
+        [dst[0], dst[128], dst[256], dst[512], dst[800], dst[1023]],
+        [-0.499, -0.366, -0.249, 0.0, 0.295, 0.492]
+    );
+
+    let (src_big, mut dst_big) = get_test_vector(128.0, Some(1024));
+    let _quant_big = quantize_roundtrip::<BlockQ2K>(src_big.as_slice(), dst_big.as_mut_slice())?;
+    compare_with_error(src_big.as_slice(), dst_big.as_slice(), 3, 6.0);
+    Ok(())
+}
+
+#[test]
+fn quantize_q4k() -> Result<()> {
+    use k_quants::BlockQ4K;
+
+    let (src, mut dst) = get_test_vector(0.5, Some(1024));
+    let _quant = quantize_roundtrip::<BlockQ4K>(src.as_slice(), dst.as_mut_slice())?;
+    compare_with_error(src.as_slice(), dst.as_slice(), 6, 0.1);
+
+    // Test some specific values
+    dst = round_vector(&dst);
```

Comment:
No need to change `dst`, `let dst = ...` should work fine if I'm not missing something.

Reply:
This wasn't fixed or am I missin something? It's not that important though but when reading the code, people would wonder why `dst` mutable rather than defined a second time.